### PR TITLE
Add support for `add_tokens()` to StableDiffusion in order to support TextualInversion

### PR DESCRIPTION
**Pull request by LukeWood**
This is done by creating a new CLIPEmbedding, manually creating a new entry in the tokenizer, and assigning the weights to the new token embedding based on a `np.mean()` call to the old weights.

My question for @mattdangerw and @fchollet:

is this a reasonable initialization strategy?  Should we randomly initialize these weights, or take a mean like this?

### CODE REVIEW

**Comment by LukeWood** [comment]
Adding @mattdangerw as a reviewer - some NLP side guidance on how to implement this would be greatly appreciated!

File: keras_cv/models/generative/stable_diffusion/clip_tokenizer.py [hunk 1]
```diff
@@ -89,18 +89,55 @@ def __init__(self, bpe_path=None):
...
+    def start_of_text(self):
+        return self.encoder["<|startoftext|>"]
+
+    def add_tokens(self, *args):
+        token = args[0]
+        if self._overridden:
+            raise ValueError("For now KerasCV can only finetune on one special token.")
+        self._overridden = True
+        self.vocab[48973] = token
+        # 48973 is the index of ðŁĺĤðŁĺĤðŁĺĤðŁĺĤðŁĺĤðŁĺĤðŁĺĤ</w>"
```
Comment (jbischof) [inline keras_cv/models/generative/stable_diffusion/clip_tokenizer.py:129]: What is the usecase for this token? If it is for classification loss like BERT the established practice is to use `0`.

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/clip_tokenizer.py:129]: Unfortunately I just looked through CLIP and there are no reserved tokens like that.

That is really unfortunate; but there ARE many garbage tokens we could throw out and replace

**Comment by LukeWood** [comment]
Matt's feedback:

- you can override the ClipEmbedding and just set the weights with the new initialized column of embeddings.  - Embeddings are just a lookup table of words -> column vectors

**Comment by LukeWood** [comment]
@mattdangerw and @fchollet this is now ready for review! 

It supports multiple tokens by creating a new lookup column in the Token Embedding, and then assigning the new row to a mean of the previous weights.  This was @fchollet 's recommendation, but I'd be curious to hear everyones thoughts on this strategy.  Perhaps this is just something I'll have to test in the guide.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 2]
```diff
@@ -111,6 +112,35 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
         self.diffusion_model.load_weights(diffusion_model_weights_fpath)
         self.decoder.load_weights(decoder_weights_fpath)
 
+    def add_tokens(self, tokens):
```
Comment (fchollet) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: This should not be on this class IMO. Time to refactor `StableDiffusion` to enable a growing set of modular use cases.

`StableDiffusion` should hold the 3 base models + a public interface. Then the actual logic should be in classes such as `TextToImage`, etc (or perhap in plain functions living somewhere like `text_to_image.py`). Something like

```python
# in class StableDiffusion
def text_to_image(self, *args, **kwargs):
   return text_to_image(self.encoder, self.decoder, self.diffusion_model, *args, **kwargs)
```

This enables growing our set of uses cases while keeping the logic modular and manageable.

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py [hunk 3]
```diff
@@ -111,6 +112,35 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
...
+        )
+        if len(old_token_weights) > 1:
+            # this should never happen, but we should warn users if something unexpected
+            # changes upstream.
+            raise ValueError("Invalid old_token_weights or old_position_weights")
+
+        old_token_weights = old_token_weights[0]
+        new_weights = np.mean(old_token_weights, axis=0)
+        new_weights = np.expand_dims(new_weights, axis=0)
+        new_weights = np.repeat(new_weights, added_tokens, axis=0)
```
Comment (fchollet) [inline keras_cv/models/generative/stable_diffusion/stable_diffusion.py:131]: You should add noise to avoid initializing multiple embeddings to the same value

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/models/generative/stable_diffusion/stable_diffusion.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: Sounds good.  I'll refactor in this pr.

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: While writing the tutorial and doing research on this I found that people tend to fetch an initialization for another similar object.

so for <my-silly-cat-token> you’d initialize with the embedding for “cat”, and for “Luke” you’d initialize with “human”

Comment (fchollet) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: That makes sense, you want an initialization mechanism that minimizes the distance the embedding will have to travel during fine-tuning.

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: So, would you suggest along with the refactor mentioned about that this method live outside of stable diffusion (perhaps in the keras.io tutorial?).  If so, then we'd have to export the sub-models from KerasCV, which might be a little ugly.

Comment (LukeWood) [reply keras_cv/models/generative/stable_diffusion/stable_diffusion.py:115]: Currently I'm thinking:

keep add_tokens() in the tokenizer, move the rest to the guide.  This will let us actually show off the weight initialization scheme via adjacent token, etc.

**Comment by LukeWood** [comment]
Ok @fchollet and @ianstenbit this is now a bit more barebones.  It basically just supports adding new tokens to the text encoder

File: keras_cv/models/generative/stable_diffusion/clip_tokenizer.py [hunk 4]
```diff
@@ -109,6 +128,22 @@ def end_of_text(self):
     def start_of_text(self):
         return self.encoder["<|startoftext|>"]
 
+    def add_tokens(self, tokens):
```
Comment (ianstenbit) [inline keras_cv/models/generative/stable_diffusion/clip_tokenizer.py:131]: Maybe this method is deserving of a bit of documentation?

Comment (ianstenbit) [reply keras_cv/models/generative/stable_diffusion/clip_tokenizer.py:131]: (In particular calling out that this is expensive and causes re-creation of the encoder/decoder)

**Review by ianstenbit — APPROVED** [review]
